File: JeMPI_Apps/JeMPI_EM_Scala/src/main/scala/org/jembi/jempi/em/InteractionEnvelop.scala
package org.jembi.jempi.em

import com.fasterxml.jackson.annotation.JsonIgnoreProperties

import scala.collection.immutable.ArraySeq

@JsonIgnoreProperties(ignoreUnknown = true)
case class InteractionEnvelop(
    contentType: String,
    tag: Option[String],
    stan: Option[String],
    interaction: Option[Interaction]
) {}

@JsonIgnoreProperties(ignoreUnknown = true)
case class Interaction(
    uniqueInteractionData: UniqueInteractionData,
    demographicData: DemographicData
)

@JsonIgnoreProperties(ignoreUnknown = true)
case class UniqueInteractionData(auxId: String)

@JsonIgnoreProperties(ignoreUnknown = true)
case class DemographicField(name: String, value: String)

@JsonIgnoreProperties(ignoreUnknown = true)
case class DemographicData(fields: ArraySeq[DemographicField])

//    givenName: String,
//    familyName: String,
//    gender: String,
//    dob: String,
//    city: String,
//    phoneNumber: String,
//    nationalId: String
//)

File: JeMPI_Apps/JeMPI_LibShared/src/main/java/org/jembi/jempi/shared/models/InteractionEnvelop.java
package org.jembi.jempi.shared.models;

import com.fasterxml.jackson.annotation.JsonInclude;

@JsonInclude(JsonInclude.Include.NON_NULL)
public record InteractionEnvelop(
      ContentType contentType,
      String tag,
      String /* System Trace Audit Number */ stan,
      Interaction interaction) {

   public enum ContentType {
      BATCH_START_SENTINEL(ContentType.MESSAGE_START_VALUE),
      BATCH_END_SENTINEL(ContentType.MESSAGE_END_VALUE),
      BATCH_INTERACTION(ContentType.MESSAGE_INTERACTION_VALUE);
      public static final int MESSAGE_START_VALUE = 1;
      public static final int MESSAGE_END_VALUE = 2;
      public static final int MESSAGE_INTERACTION_VALUE = 3;

      public final int type;

      ContentType(final int type_) {
         this.type = type_;
      }
   }

}


